umbraco/Umbraco-CMS
# Package telemetry

This adds the functionality to send telemetry information about installed packages.

The `TelemetryService` uses the `IManifestParser` to look for `package.manifests` in the App_Plugins folder. By default the names and versions of the installed packages will be sent, provided that the package has specified a version in the new `version` property  in the `package.manifest`. For the name of the package, the name of the folder will be used, unless the package creator has specified a different name with the `name` property in the `package.manifest`

If a package creator does not wish for there to be sent telemetry information about their package at all they can set `allowPackageTelemetry` to false in their `package.manifest` and then the package will be skipped entirely.

If a site developer wishes to restrict the information sent about the installed package they can set the global setting `RestrictPackageTelemetry` to true in their appsettings.json, and then the version of the installed packages will be omitted: 

```json
"CMS": {
  "Global": {
    "RestrictPackageTelemetry": true
  }
}
```

## Testing

1. Test pass 
2. Try and create some packages in the `App_Plugins` folder, there doesn't need to be any content, just a folder, and a package.maniest
3. Ensure that if you add a "name" property it will be used for the package name, otherwise the folder name should be used
4. Ensure that "version" property is used if there is any, otherwise it should just be an empty string
5. Ensure that the package is completely ignored if the "allowPackageTelemetry" is set to false
6. Try and set the `RestrictPackageTelemetry` global setting to true, and ensure that no package versions are sent regardless

When testing this you can either put a breakpoint in `ReportSiteTask` and look at the `TelemetryReportData` that the telemetry service returns when it runs, or if you want to go the full mile you can set up a server to retrieve the request that is sent, and ammend the `s_httpClient.BaseAddress` to instead use a URI of your server, I used a simple Flask server with python for this. You might want to reduce the delay so you don't have to wait a full minute every time.

Example `package.manifest` with the new properties:

```json
{
    "allowPackageTelemetry": false,
    "name": "Do Not Track",
    "version": "2.1.5"
}
```

Flask test server:

```python
import json

from flask import Flask, app
from flask.globals import request

app = Flask("JsonDumpServer")


@app.route("/installs/", methods=['POST'])
def dump_json():
    print("Received request\n\n")
    print(f"{json.dumps(request.get_json(), indent=2)}\n\n")
    return "Request received"


app.config['ENV'] = 'development'
app.config['TESTING'] = True
app.run()

```
